src/Dofs/DofHandler.jl

```diff
@@ -0,0 +1,250 @@
+
+# Stores the dofs for the cell with number `i` into the vector `global_dofs`
+function celldofs!(global_dofs::Vector{Int}, dh::DofHandler, i::Int)
+    @assert isclosed(dh)
+    @assert length(global_dofs) == ndofs_per_cell(dh)
+    @inbounds for j in 1:ndofs_per_cell(dh)
+        global_dofs[j] = dh.dofs_cells[j, i]
+    end
+    return global_dofs
+end
+
+# Add a collection of fields
+function push!(dh::DofHandler, names::Vector{Symbol}, dims)
+    @assert length(names) == length(dims)
+    for i in 1:length(names)
+        push!(dh, names[i], dims[i])
+    end
+end
+
+# Add a field to the dofhandler ex `push!(dh, :u, 3)`
+function push!(dh::DofHandler, name::Symbol, dim::Int)
+    @assert !isclosed(dh)
+    if name in dh.field_names
+        error("duplicate field name")
+    end
+    push!(dh.field_names, name)
+    push!(dh.dof_dims, dim)
+    append!(dh.dofs_vec, length(dh.dofs_vec)+1:length(dh.dofs_vec) +  dim * getnnodes(dh.grid))
+    return dh
+end
+
+# Computes the number of dofs from which the field starts data
+# For example [ux, uy, uz, T] --> dof_offset(dh, :temperature) = 4
+function dof_offset(dh::DofHandler, field_name::Symbol)
+    offset = 0
+    i = 0
+    for name in dh.field_names
+        i += 1
+        if name == field_name
+            return offset
+        else
+            offset += dh.dof_dims[i]
+        end
+    end
+    error("unexisting field name $field_name among $(dh.field_names)")
+end
+
+function ndim(dh::DofHandler, field_name::Symbol)
+    i = 0
+    for name in dh.field_names
+        i += 1
+        if name == field_name
+            return dh.dof_dims[i]
+        end
+    end
+    error("unexisting field name $field_name among $(dh.field_names)")
+end
+
+function close!(dh::DofHandler)
+    @assert !isclosed(dh)
+    dh.dofs_nodes = reshape(dh.dofs_vec, (length(dh.dofs_vec) ÷ getnnodes(dh.grid), getnnodes(dh.grid)))
+    add_element_dofs!(dh)
+    dh.closed[] = true
+    return dh
+end
+
+getnvertices{dim, N, M}(::Type{JuAFEM.Cell{dim, N, M}}) = N
+
+# Computes the "edof"-matrix
+function add_element_dofs!(dh::DofHandler)
+    n_elements = getncells(dh.grid)
+    n_vertices = getnvertices(getcelltype(dh.grid))
+    element_dofs = Int[]
+    ndofs = size(dh.dofs_nodes, 1)
+    for element in 1:n_elements
+        offset = 0
+        for dim_doftype in dh.dof_dims
+            for node in getcells(dh.grid, element).nodes
+                for j in 1:dim_doftype
+                    push!(element_dofs, dh.dofs_nodes[offset + j, node])
+                end
+            end
+            offset += dim_doftype
+        end
+    end
+    dh.dofs_cells = reshape(element_dofs, (ndofs * n_vertices, n_elements))
+end
+
+# Creates a sparsity pattern from the dofs in a dofhandler.
+# Returns a sparse matrix with the correct pattern.
+function create_sparsity_pattern(dh::DofHandler)
+    grid = dh.grid
+    n = ndofs_per_cell(dh)
+    N = n^2 * getncells(dh.grid)
+    I = Int[]
+    J = Int[]
+    sizehint!(I, N)
```

Comment:
I think the speed should be exactly the same but allocating up front is probably cleaner.